### DISCUSSION TITLE
Twister fails for STM32H743 when running on pipeline but locally it passes

### DISCUSSION THREAD

**Discussion by harshrmistry-eaton** [zephyrproject-rtos/zephyr#76572] [General]
Hi, I have a machine where our pipeline tests are setup. When I trigger a twister run on the machine, it fails on the pipeline with Device issue (Flash error?). Locally it passes on my machine without any issues. I am running twister test for Nucleo-H743ZI2 board. Below is the snapshot of the error:
![twister_error](https://github.com/user-attachments/assets/646a1024-1e97-4ca4-b63f-9e3f5bec6e76)

I am using Zephyr 3.6.

Below is the command used in the pipeline:
`python scripts/twister --device-testing --device-serial COM3 -p nucleo_h743zi -T <path to the root directory for my examples> -O <path to a directory> -c`

**Comment by FRASTM** [comment]
Hi @harshrmistry-eaton 
I would recommend to update the stlink of your nucleo_h743zi board with the cube programmer.
Then you could check if your twister and your west flash are using the same runner.

**Reply by harshrmistry-eaton** [reply]
I will try that and update.

What seems strange is locally when I run the twister command with the Nucleo H743ZI2 board connected, it runs fine. No error is shown. Even though I didn't update the stlink version. Also, there is one particular agent machine where it fails all the time.

I have also tried adding the -vv option and the --west-flash="--runner=openocd" options but it doesn't work.

**Reply by harshrmistry-eaton** [reply]
@FRASTM I tried upgrading the stlink version on one of the agent machines. The twister build is passing. Now checking other machines. 

Thanks a lot for the suggestion.

**Reply by harshrmistry-eaton** [reply]
Also, I wanted to know how can we check stlink version using a command? Is there some way to upgrade the stlink version on the board using command line?

**Reply by harshrmistry-eaton** [reply]
@FRASTM If you could please update on the above questions then it would help.

**Reply by FRASTM** [reply]
Sorry I do not know CLI for [STLinkUpgrade](https://www.st.com/en/development-tools/stsw-link007.html)

**Reply by harshrmistry-eaton** [reply]
@FRASTM I tried running twister build on a machine without stlink upgrade and it passed. Is there a minimum requirement for twister to run on stlink version? What more can we look into to find the root cause?

**Reply by FRASTM** [reply]
I do not know much about the tools. The version of the STlinkupgrade  and the stm32Cubeprogrammer probably have  dependency.
What is your STM32CubeProgrammer version ?

**Reply by FRASTM** [reply]
I found this:
with [STLinkUpdate](https://www.st.com/en/development-tools/stsw-link007.html) you can use the following command line:

- check the current fw version on your board  : `java -jar STLinkUpgrade.jar -checkVer`
- update the fw version of your board (if a more recent one is detected) : `java -jar STLinkUpgrade.jar -update`

**Reply by harshrmistry-eaton** [reply]
@FRASTM The issue is quite random. It doesn't happen locally but only on pipeline. But our machines are identical. The only difference is some have Windows 10 and some have Windows 11. The st-link versions were different for the respective boards connected to our machines. The machines don't have STM32CubeProgrammer installed. Only one machine has ST-LINK Utility installed since that's a debug machine.

It would be helpful if someone from ST could help with the minimum st-link firmware version required for twister to run. We are currently on Zephyr 3.6.0. Also, does twister use openocd while flashing? Is it necessary to have STM32CubeProgrammer installed on the machines?

**Reply by FRASTM** [reply]
If the runner is openocd (that's the case with board.cmake), flashing is done with openocd,  no need for stm32CubeProgrammer 
However the openocd is detecting the STLink which is present on the target board. For example:
```
Open On-Chip Debugger 0.11.0+dev-00726-ge2b6f5655 (2023-09-22-14:29)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
init
Info : STLINK V3J13M4 (API v3) VID:PID 0483:374E
```

**Reply by harshrmistry-eaton** [reply]
@FRASTM We upgraded all H7 boards to V3J15M6 version. We again started facing this issue.

I have posted in ST forum as well but haven't got any concrete response. Please help regarding this.

**Reply by FRASTM** [reply]
I guess your twister has openocd as runner, What are the build.log and device.log  when failing ?

Can you also try with pyocd as runner for twister scripting
